hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java

```diff
@@ -169,19 +169,24 @@ private RexNode convertCast(SqlCall call, Blackboard blackboard) {
 
     /**
      * This method overcomes a bug in Apache Calcite that ignores previously resolved return types of the expression
-     * and instead attempts to infer them again using different logic. Without this fix, we will get type resolution
-     * errors after SQL-to-rel conversion.
+     * and instead attempts to infer them again using a different logic. Without this fix, we will get type resolution
+     * errors after a SQL-to-rel conversion.
      * <p>
      * The method relies on the fact that all operators use {@link HazelcastReturnTypeInference} as a top-level return type
      * inference method.
      * <ul>
-     *     <li>When a call node is observed for the first time, get it's return type and save it to a thread local variable</li>
-     *     <li>Then delegate back to original converter code</li>
-     *     <li>When converter attempts to resolve the return type of a call, he will get the previously saved type from
-     *     the thread-local variable</li>
+     *     <li>When a call node is observed for the first time, get its return type and save it to a thread-local variable
+     *     <li>Then delegate back to original converter code
+     *     <li>When converter attempts to resolve the return type of a call, it will get the previously saved type from
+     *     the thread-local variable
      * </ul>
      */
     private RexNode convertCall(SqlNode node, Blackboard blackboard) {
+        // ignore DEFAULT (used for default function arguments). It doesn't support getValidatedNodeType()
+        if (node.getKind() == SqlKind.DEFAULT) {
```

Comment:
What is the result of sql-to-rex conversion of the `SqlNode` with the `SqlKind.DEFAULT`?

Reply:
It's this: 
![image](https://user-images.githubusercontent.com/9109877/103290754-8f487f00-49ea-11eb-8719-950a7537aa07.png)

Reply:
Do we understand exactly why `getValidatedNodeType` is not supported here? Does it mean that the `DEFAULT` node is not validated? Because if it has been validated before, it should be saved in the validator's node-to-type map.

Reply:
The exception it throws is this:
```
Caused by: java.lang.UnsupportedOperationException: class org.apache.calcite.sql.SqlBasicCall: DEFAULT
	at org.apache.calcite.util.Util.needToImplement(Util.java:967)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.getValidatedNodeType(SqlValidatorImpl.java:1701)
	at com.hazelcast.sql.impl.calcite.HazelcastSqlToRelConverter.convertCall(HazelcastSqlToRelConverter.java:193)
```
Btw, the node before translation is this:
![image](https://user-images.githubusercontent.com/9109877/104206019-5c14df80-542f-11eb-9805-dc28dc527269.png)

I debugged it a bit more and the reason might be this: We have a function with 4 arguments. But we only pass 3, the remaining one is default. In `SqlCallBinding.permutedCall()` it matches the actual operands with defined, and add the 4th argument to `SqlCallBinding.DEFAULT_CALL`. Therefore, this SqlNode is not present in the tree when validating, but is added when converting. I don't think you can pass a DEFAULT value explicitly, so it can be only created this way.

Reply:
Thank you for the explanation. I am ok with the change then. Maybe be just a minor thought: it would be great to extend the comment with your findings. So that we always remember, that the `DEFAULT` operator is treated in a special way because it is created inside the `SqlCallBinding.permutedCall()`, and is not validated as other nodes.